proposals/user-scripts-api.md

```diff
@@ -0,0 +1,165 @@
+
+## Proposal
+
+### New Namespace
+
+User scripting related features will be exposed in a new API namespace, tentatively named `userScripts`.  The proposal authors favor the use of a new namespace for several reasons.
+
+1. **Better set developer expectations.**  The clear separation between user and content scripts will reduce confusion for developers for which API methods to use and what the capabilities / restrictions of each are.  This naming also more clearly communicates that this capability is not meant as a general purpose way to inject arbitrary scripts.
+
+2. **Stronger enforcement of remotely hosted code abuse.**  A distinct namespace allows people to more clearly see where user scripts features are being used and more easily spot abuse patterns.
+
+3. **Easier engineering implementation.**  Browser vendors should be able to restrict the execution world and different API methods behind features.
+
+4. **Introduce a user script world with custom CSP.**  Allow user scripts to opt-in to a more secure world that is only available in this namespace.
+
+### API Schema
+
+#### Types
+
+```
+dictionary RegisteredUserScript {
+  boolean? allFrames;
+  string code;
+  string[]? excludeMatches;
+  string id;
+  string[]? matches;
+  RunAt runAt;
+  // Allows `USER_SCRIPT` (default) and `MAIN`
+  // and returns error for `ISOLATED`.
+  ExecutionWorld? world;
+  string[]? includeGlobs;
+  string[]? excludeGlobs;
+}
+dictionary UserScriptFilter {
+  string[]? ids;
+}
+```
+
+#### Methods
+
+```
+browser.userScripts.register(
+  scripts: RegisteredUserScript[],
+  callback?: function
+)
+browser.userScripts.unregister(
+  filter?: UserScriptFilter[],
+  callback?: function
+)
+browser.userScripts.getScripts(
+  filter?: UserScriptFilter[],
+  callback?: function
+)
+browser.userScripts.update(
+  scripts?: RegisteredUserScript[],
+  callback?: function
+)
+```
+
+### Requirements
+
+#### A. A mechanism to execute code in the main world
+
+- User scripts can be registered in the `MAIN` world.
+- Extension can customize the `USER_SCRIPT` world's CSP to inject a script tag into the host page (expanded in requirement B)
+
+#### B. The ability to execute code (with a separate CSP) in a world different from the main world and the extension's isolated world
+
+##### `USER_SCRIPT` World
+
+- Isolated from the web page (similar to other isolated worlds), but will be un-permissioned.  It will not have access to any extension APIs or cross-origin exceptions (these don't exist in Chrome, but do in other browsers).
+- Exempt from the page's CSP (see [Relax CSP](#relax-csp)).
+- Share DOM with the web page.  Code from both worlds cannot directly interact with each other, except through DOM APIs.
+- Can communicate with different JS worlds via [`window.postMessage()`](https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage).  A dedicated API to communicate between worlds is being considered as part of future work.
```

Comment:
[`window.postMessage`](https://developer.mozilla.org/en-US/docs/Web/API/Window/postMessage) has the argument `targetOrigin`.
Maybe it can be made use of by using that argument to specify the exact userscript.

Using an Event is also a bad idea:

1. It's synchronous to pass data
2. There's no way to respond to a request. Instead it's close to being in a UDP message passing system. 
3.  There's no delivery guarantees (well... postmessage() also doesn't but it's a bad thing of both) 
4. There's no way to pass data without serialization (no transferrable content allowed)
   1.  This can be horrible for `XMLHTTPRequest` or `fetch` performance

If you want to present an alternative, present [`MessagePort`](https://developer.mozilla.org/en-US/docs/Web/API/MessagePort)!

Assynchronous message passing which has transferrable and the destination is the owner of the other side of the port.
Conceptually simple that solves all of these issues except this one:

How to deliver the port to each side?

Reply:
> In defence of the current proposal

The proposal doesn't provide secure way of using any kind of messaging so you're not really defending it. Even with postMessage origin, we can't target userscripts in the main world as their origin is not unique. Also, when such a main world's userscript sends a postMessage to the extension's content script in the isolated world, the extension can't be sure that the message wasn't faked by a hostile unrelated script.

> It's synchronous

Userscript managers such as Violentmonkey/Tampermonkey still need it to implement the classic Greasemonkey API, which is used by the majority of userscripts, many of which will never be updated. There are many other cases when a synchronous response is required, which is why CustomEvent exists on equal terms with asynchronous methods.

There's nothing bad about synchronous communication because the actual internal work under the hood is identical for both asynchronous and synchronous methods, the only reason why postMessage or MessageChannel are asynchronous is because they are designed for cross-origin communication where environments use a different OS thread/process. In the same JS environment postMessage may be effectively slower than CustomEvent due to the former being much more complicated inside and the requirement to deliver the message in a separate event loop task per the specification.

> There's no way to respond to a request

Simply send a message back immediately (it's synchronous as well) or add an id to every message.

> There's no delivery guarantees

The extension has added a listener, so there is a guarantee.

> There's no way to pass data without serialization

Indeed - in case you mean _structured cloning_ - but passing a Blob of any size takes just 1ms via CustomEvent or postMessage because blobs are pointers to data not the data itself. To pass other binary types of data we can wrap it in a Blob, probably.

> How to deliver the port to each side?

That's the essence of the problem: the API should provide a secure method of communication.